### PR TITLE
CAF-3518: JavaScript post-processing

### PR DESCRIPTION
**Pull request by kploch**
https://jira.autonomy.com/browse/CAF-3518

### CODE REVIEW

File: worker-document-interface/src/main/java/com/hpe/caf/worker/document/model/ResponseOptions.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+ */
+package com.hpe.caf.worker.document.model;
+
+import java.util.Map;
+
+/**
+ * Class containing additional worker response properties.
+ * This class is immutable.
+ */
+public class ResponseOptions
```
Comment (dermot-hardy) [inline worker-document-interface/src/main/java/com/hpe/caf/worker/document/model/ResponseOptions.java]: - Sorry but this should be an interface and if it wants to be in this package with all the other model objects then it needs to extend `DocumentWorkerObject`.
- Sorry but it should also have a `getTask()` to get back to the task that it is a response for.
- My first impression is that we should just call this `Response`.  I'm not sure that the `Options` suffix is really adding anything.

File: worker-document-interface/src/main/java/com/hpe/caf/worker/document/model/ResponseOptions.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    {
+        return queueName;
+    }
+
+    /**
+     * Gets the custom data.
+     *
+     * @return The custom data map.
+     */
+    public Map<String, String> getCustomData()
```
Comment (dermot-hardy) [inline worker-document-interface/src/main/java/com/hpe/caf/worker/document/model/ResponseOptions.java]: Why do we need to expose this?  If it must be exposed then unfortunately I think it should be exposed as a custom collection type in the same way as Fields, Failures, Subdocuments, FieldValues.

File: worker-document-interface/src/main/java/com/hpe/caf/worker/document/model/Task.java [hunk 3]
```diff
@@ -64,4 +65,17 @@
      * @see Application#getService(Class) Application.getService()
      */
     <S> S getService(Class<S> service);
+
+    /**
+     * Allows to set additional response options for this task.
+     * @param queueName Queue name to post response message to.
+     * @param customData Custom data to set on a response message.
+     */
+    void setResponseOptions(String queueName, Map<String, String> customData);
```
Comment (dermot-hardy) [inline worker-document-interface/src/main/java/com/hpe/caf/worker/document/model/Task.java]: This method made sense by itself, but there doesn't seem to be much point in it given that you've had to introduce the Response object.  Why not just add a `setQueueName()` etc. to it?

File: worker-document-interface/src/main/java/com/hpe/caf/worker/document/model/Task.java [hunk 4]
```diff
@@ -64,4 +65,17 @@
...
+
+    /**
+     * Retrieves additional response options for this task.
+     * @return Response options if there are any or {@code null}.
```
Comment (dermot-hardy) [inline worker-document-interface/src/main/java/com/hpe/caf/worker/document/model/Task.java]: Why would this ever be `null`?  I assume you were thinking that this should only be available if `setResponseOptions()` has been called.  It would be better to just make it always available and have the setter options available on it.

**Review by dermot-hardy — CHANGES_REQUESTED** [review]
I've only commented on the `worker-document-interface` changes.  I'll take a closer look at the implementation on Monday (or over the weekend if I get a chance).

File: worker-document-interface/src/main/java/com/hpe/caf/worker/document/model/ResponseOptions.java (hunk 2 above)
Comment (kploch) [reply worker-document-interface/src/main/java/com/hpe/caf/worker/document/model/ResponseOptions.java]: Dermot, while I understand the object model consistency preference (having a parent etc. for the ResponseOptions), I don't follow why you would introduce a custom collection type here. CustomData type is a map.

File: worker-document-interface/src/main/java/com/hpe/caf/worker/document/model/ResponseOptions.java (hunk 1 above)
Comment (kploch) [reply worker-document-interface/src/main/java/com/hpe/caf/worker/document/model/ResponseOptions.java]: Re naming - Response vs ResponseOptions - IMO Response name is too confusing. This class doesn't have full Response data. It only has (additional) options for the response.

Comment (dermot-hardy) [reply worker-document-interface/src/main/java/com/hpe/caf/worker/document/model/ResponseOptions.java]: We may add more stuff to it in the future about how the response is formed.

File: worker-document-interface/src/main/java/com/hpe/caf/worker/document/model/ResponseOptions.java (hunk 2 above)
Comment (dermot-hardy) [reply worker-document-interface/src/main/java/com/hpe/caf/worker/document/model/ResponseOptions.java]: You would probably have said the same about the other collection types - fields is just a map, failures is just a list of failures, etc.  Having it as a custom collection means we can control what functionality we want to make available and can change the implementation if necessary.

File: worker-document-interface/src/main/java/com/hpe/caf/worker/document/model/ResponseOptions.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+ * limitations under the License.
+ */
+package com.hpe.caf.worker.document.model;
+
+import java.util.Map;
+
+/**
+ * Interface for additional worker response properties.
+ */
+public interface ResponseOptions extends DocumentWorkerObject
```
Comment (dermot-hardy) [inline worker-document-interface/src/main/java/com/hpe/caf/worker/document/model/ResponseOptions.java]: I'm sorry but I still don't understand why this isn't just `Response`.  Is it that it doesn't represent the final immutable response but rather just something that effects that final response, in much the same way that modifying the document does?

File: worker-document-interface/src/main/java/com/hpe/caf/worker/document/model/ResponseOptions.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+{
+    /**
+     * Gets the queue name that will be used when a worker sends response message.
+     * @return The queue name.
```
Comment (dermot-hardy) [inline worker-document-interface/src/main/java/com/hpe/caf/worker/document/model/ResponseOptions.java]: I would never make this comment for any other module but in `worker-document-interface` the JavaDoc guidelines are followed religiously - http://www.oracle.com/technetwork/articles/java/index-137868.html - `@return` and `@param` lines should be phrases rather than sentences (i.e. shouldn't start with capital letters or end in full stops).

File: worker-document-interface/src/main/java/com/hpe/caf/worker/document/model/Task.java [hunk 7]
```diff
@@ -64,4 +64,11 @@
      * @see Application#getService(Class) Application.getService()
      */
     <S> S getService(Class<S> service);
+
+    /**
+     * Retrieves additional response options for this task.
+     * @return Response options if there are any or {@code null}.
```
Comment (dermot-hardy) [inline worker-document-interface/src/main/java/com/hpe/caf/worker/document/model/Task.java]: This JavaDoc says `null` may be returned but I see the `@Nonnull` annotation is applied.  I hope the JavaDoc is wrong because that will make the method easier to use.

File: worker-document-interface/src/main/java/com/hpe/caf/worker/document/model/ResponseOptions.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+import java.util.Map;
+
+/**
+ * Interface for additional worker response properties.
+ */
+public interface ResponseOptions extends DocumentWorkerObject
+{
+    /**
+     * Gets the queue name that will be used when a worker sends response message.
```
Comment (dermot-hardy) [inline worker-document-interface/src/main/java/com/hpe/caf/worker/document/model/ResponseOptions.java]: I think we could describe here how the queue is decided if it hasn't been explicitly set by the `setQueueName()` method.

File: worker-document-shared/src/main/java/com/hpe/caf/worker/document/DocumentWorkerConstants.java [hunk 9]
```diff
@@ -26,6 +26,12 @@
     public static final String DOCUMENT_TASK_NAME = "DocumentWorkerTask";
     public static final int DOCUMENT_TASK_API_VER = 1;
 
+    /**
+     * Custom data setting for the post processing JavaScript.
+     * The setting value is expected to be a data store reference.
+     */
+    public static final String POST_PROCESSING_SCRIPT_CUSTOM_DATA = "postProcessingScript";
```
Comment (dermot-hardy) [inline worker-document-shared/src/main/java/com/hpe/caf/worker/document/DocumentWorkerConstants.java:33]: As discussed, I'd prefer if this was a full member of `DocumentWorkerDocumentTask` rather than being in custom data:
 - the code that acts on it is in the framework rather than the (custom) worker code
 - it is only applicable to this type of task (and not the field enrichment type) (at least right now)

File: worker-document/src/main/java/com/hpe/caf/worker/document/DocumentMessageProcessor.java [hunk 10]
```diff
@@ -81,6 +81,7 @@ public final WorkerResponse doWork() throws InterruptedException, TaskRejectedEx
         try {
             processDocument(document);
         } catch (DocumentWorkerTransientException dwte) {
+
```
Comment (dermot-hardy) [inline worker-document/src/main/java/com/hpe/caf/worker/document/DocumentMessageProcessor.java]: Pointless whitespace change

File: worker-document/src/main/java/com/hpe/caf/worker/document/DocumentPostProcessor.java [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+ * An interface a document post-processor.
+ * The {@code postProcessDocument} will be called when core processing of a document is finished.
+ */
+public interface DocumentPostProcessor
+{
+    /**
+     * Method called after core document processing by a worker is finished.
+     * @param document A document processed by a worker.
+     */
+    void postProcessDocument(final Document document) throws PostProcessingFailedException;
```
Comment (dermot-hardy) [inline worker-document/src/main/java/com/hpe/caf/worker/document/DocumentPostProcessor.java]: This is generating a JavaDoc warning during the build:
```
Javadoc Warnings
DocumentPostProcessor.java:32: warning: no @throws for PostProcessingFailedException
void postProcessDocument(final Document document) throws PostProcessingFailedException;
```

File: worker-document/pom.xml [hunk 12]
```diff
@@ -49,6 +49,10 @@
             <artifactId>commons-io</artifactId>
         </dependency>
         <dependency>
+            <groupId>com.google.guava</groupId>
+            <artifactId>guava</artifactId>
```
Comment (dermot-hardy) [inline worker-document/pom.xml]: Is it really worth adding a dependency just for `Strings.isNullOrEmpty()`?  I hate that Java makes me worry about things like this.  I know I was tempted to use it in other places but just went with `== null || .isEmpty()`.

File: worker-document/src/main/java/com/hpe/caf/worker/document/DocumentWorkerFactoryProvider.java [hunk 13]
```diff
@@ -42,12 +42,15 @@
...
+        final DocumentPostProcessorFactory postProcessorFactory = new DocumentPostProcessorFactory();
         // Construct the application object
-        final ApplicationImpl application = new ApplicationImpl(configSource, dataStore, codec);
+        final ApplicationImpl application = new ApplicationImpl(configSource, dataStore, codec, postProcessorFactory);
 
         // Construct the DocumentWorker implementation object
         final DocumentWorker documentWorker = createDocumentWorker(application);
 
+
+
```
Comment (dermot-hardy) [inline worker-document/src/main/java/com/hpe/caf/worker/document/DocumentWorkerFactoryProvider.java]: Pointless whitespace change

File: worker-document/src/main/java/com/hpe/caf/worker/document/DocumentPostProcessorFactory.java [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+        }
+        return null;
+    }
+
+    private static String retrieveScript(final String reference, final Document document) throws TaskRejectedException
+    {
+        final DataStore dataStore = document.getApplication().getService(DataStore.class);
+
+        try (InputStream stream = dataStore.retrieve(reference)) {
+            return IOUtils.toString(stream);
```
Comment (dermot-hardy) [inline worker-document/src/main/java/com/hpe/caf/worker/document/DocumentPostProcessorFactory.java]: This is currently platform specific.  It should be passing `StandardCharsets.UTF_8` as a second parameter.

File: worker-document/src/main/java/com/hpe/caf/worker/document/DocumentPostProcessorFactory.java [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+            return new JavaScriptDocumentPostProcessor(postProcessingScript);
+        }
+        return null;
+    }
+
+    private static String retrieveScript(final String reference, final Document document) throws TaskRejectedException
+    {
+        final DataStore dataStore = document.getApplication().getService(DataStore.class);
+
+        try (InputStream stream = dataStore.retrieve(reference)) {
```
Comment (dermot-hardy) [inline worker-document/src/main/java/com/hpe/caf/worker/document/DocumentPostProcessorFactory.java]: Should be marked `final`.

File: worker-document/src/main/java/com/hpe/caf/worker/document/DocumentPostProcessorFactory.java [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+ * Responsible for creating of {@link DocumentPostProcessor}.
+ * At the moment, this class will create a {@link JavaScriptDocumentPostProcessor} if
+ * postProcessingScript custom data is provided, otherwise {@code null}.
+ */
+public class DocumentPostProcessorFactory
+{
+    private static final Logger LOG = LoggerFactory.getLogger(DocumentPostProcessorFactory.class);
+    private final Object syncObj = new Object();
+
+    private final HashMap<String, String> cachedScripts = new HashMap<>();
```
Comment (dermot-hardy) [inline worker-document/src/main/java/com/hpe/caf/worker/document/DocumentPostProcessorFactory.java:39]: Do we have a JIRA for replacing this with a cache which has some sort of expiry policy?  Obviously this one will just keep growing.

File: worker-document-testing-unit/src/main/java/com/hpe/caf/worker/document/testing/DocumentBuilder.java [hunk 17]
```diff
@@ -146,8 +146,9 @@ public Document build() throws WorkerException
             services = TestServices.createDefault();
         }
 
+        // Post processor is not a part of document. Factory intentionally set to null in the ApplicationImpl constructor.
```
Comment (dermot-hardy) [inline worker-document-testing-unit/src/main/java/com/hpe/caf/worker/document/testing/DocumentBuilder.java]: What is the point of this complication?  Why not just create the `DocumentPostProcessorFactory` object inside  `ApplicationImpl`?  Do you want to be able to provide a subclass of `DocumentPostProcessorFactory` to `ApplicationImpl` in the future?  It would be simpler if the `DocumentPostProcessorFactory` object was just constructed inside `ApplicationImpl` without any need to change this class or `DocumentWorkerFactoryProvider`.  Constructing a `DocumentPostProcessorFactory` instance outside of `ApplicationImpl` and then passing it in without there being any need for multiple implementations of it is a premature abstraction.  It can easily get out of control and make the framework hard to follow.

File: worker-document/src/main/java/com/hpe/caf/worker/document/impl/ApplicationImpl.java [hunk 18]
```diff
@@ -39,7 +40,7 @@
     private final String successQueue;
     private final String failureQueue;
 
-    public ApplicationImpl(final ConfigurationSource configSource, final DataStore dataStore, final Codec codec)
+    public ApplicationImpl(final ConfigurationSource configSource, final DataStore dataStore, final Codec codec, DocumentPostProcessorFactory postProcessorFactory)
```
Comment (dermot-hardy) [inline worker-document/src/main/java/com/hpe/caf/worker/document/impl/ApplicationImpl.java]: I've commented elsewhere that I don't see the need to construct this at the outermost layer and pass it through, but if it is to be passed through then it should be marked `final` and the whole line should not go over the 138-character guideline.

File: worker-document/src/main/java/com/hpe/caf/worker/document/impl/ApplicationImpl.java [hunk 19]
```diff
@@ -48,7 +49,7 @@ public ApplicationImpl(final ConfigurationSource configSource, final DataStore d
         this.codec = Objects.requireNonNull(codec);
         this.configuration = getConfiguration(configSource);
         this.batchSizeController = new BatchSizeControllerImpl(this, configuration);
-        this.inputMessageProcessor = new InputMessageProcessorImpl(this, configuration.getInputMessageProcessing());
+        this.inputMessageProcessor = new InputMessageProcessorImpl(this, postProcessorFactory, configuration.getInputMessageProcessing());
```
Comment (dermot-hardy) [inline worker-document/src/main/java/com/hpe/caf/worker/document/impl/ApplicationImpl.java]: Rather than passing this to `InputMessageProcessorImpl`, and then into the `DocumentTask` constructors from there, just create a `getPostProcessorFactory()` accessor.  That way you won't need to change `InputMessageProcessorImpl` at all.  The `ApplicationImpl` object is already passed to `DocumentTask` and stored in the `application` instance variable.

File: worker-document/src/main/java/com/hpe/caf/worker/document/tasks/DocumentTask.java [hunk 20]
```diff
@@ -26,43 +29,51 @@
...
 
     private DocumentTask(
-        final ApplicationImpl application,
-        final WorkerTaskData workerTask,
-        final DocumentWorkerDocumentTask documentTask
-    ) throws InvalidChangeLogException
+            final ApplicationImpl application,
+            final WorkerTaskData workerTask,
+            final DocumentWorkerDocumentTask documentTask,
+            DocumentPostProcessorFactory postProcessorFactory) throws InvalidChangeLogException, TaskRejectedException
```
Comment (dermot-hardy) [inline worker-document/src/main/java/com/hpe/caf/worker/document/tasks/DocumentTask.java]: As discussed in another comment, you wouldn't need to pass `DocumentPostProcessorFactory` here if you added a `getPostProcessorFactory()` accessor to `ApplicationImpl`.

File: worker-document/src/main/java/com/hpe/caf/worker/document/tasks/DocumentTask.java [hunk 21]
```diff
@@ -123,8 +155,8 @@ protected WorkerResponse createWorkerResponseImpl()
     protected WorkerResponse handleGeneralFailureImpl(final Throwable failure)
     {
         document.getFailures().add(failure.getClass().getName(),
-                failure.getLocalizedMessage(),
-                failure);
+                                   failure.getLocalizedMessage(),
+                                   failure);
```
Comment (dermot-hardy) [inline worker-document/src/main/java/com/hpe/caf/worker/document/tasks/DocumentTask.java]: Pointless whitespace change

File: worker-document/src/main/java/com/hpe/caf/worker/document/tasks/DocumentTask.java [hunk 22]
```diff
@@ -102,10 +123,21 @@ protected WorkerResponse createWorkerResponseImpl()
         documentWorkerResult.document = documentTask.document;
         documentWorkerResult.changeLog = changeLog;
 
+        ResponseOptions responseOptions = getResponseOptions();
```
Comment (dermot-hardy) [inline worker-document/src/main/java/com/hpe/caf/worker/document/tasks/DocumentTask.java]: Should be `final`.

File: worker-document/src/main/java/com/hpe/caf/worker/document/tasks/DocumentTask.java [hunk 23]
```diff
@@ -102,10 +123,21 @@ protected WorkerResponse createWorkerResponseImpl()
...
-        final String outputQueue = ChangeLogFunctions.hasFailures(changes)
-            ? application.getFailureQueue()
-            : application.getSuccessQueue();
+        // TODO: ResponseOptions queue name will override the queue set below. This means that failure queue will
+        // not be used in case of failures. This is correct behaviour but we might want to add ability
+        // to set a failure queue on response options.
+        String outputQueue = ChangeLogFunctions.hasFailures(changes) ? application.getFailureQueue()
+                : application.getSuccessQueue();
+
+        if (responseOptions.getQueueName() != null) {
```
Comment (dermot-hardy) [inline worker-document/src/main/java/com/hpe/caf/worker/document/tasks/DocumentTask.java]: - The JavaDoc for `getQueueName()` reads "Gets the queue name that will be used when a worker sends response message."  I had thought from the description that it would return the success or failure queue if the queue wasn't explicitly set, but I see now from how it is being used that it only returns if the queue has been explicitly set.  I'm not sure I like the `getQueueName()` method name anymore.  Perhaps it should be `getQueueNameOverride()` or something like that?  Perhaps that is why you are so keen that the class is called `ResponseOptions` rather than the more definitive `Response`?
- Since the explicitly set queue name takes priority why not check it first, avoiding the need to call `ChangeLogFunctions.hasFailures()` if it is not necessary.

File: worker-document/src/main/java/com/hpe/caf/worker/document/tasks/AbstractTask.java [hunk 24]
```diff
@@ -93,4 +104,58 @@ public final WorkerResponse handleGeneralFailure(final Throwable failure)
 
     @Nonnull
     protected abstract WorkerResponse handleGeneralFailureImpl(final Throwable failure);
+
+    private static class ResponseOptionsImpl implements ResponseOptions
```
Comment (dermot-hardy) [inline worker-document/src/main/java/com/hpe/caf/worker/document/tasks/AbstractTask.java]: - I don't see the advantage of this being a nested class.  Why not just create a full proper class for it alongside the other object model implementations?
- Why not extend `DocumentWorkerObjectImpl` to implement `getApplication()`?  It doesn't save you much but all the other object model implementations use it.

File: worker-document/src/main/java/com/hpe/caf/worker/document/tasks/AbstractTask.java [hunk 25]
```diff
@@ -93,4 +104,58 @@ public final WorkerResponse handleGeneralFailure(final Throwable failure)
...
+        private final Task parentTask;
+        private final Application application;
+
+        private ResponseOptionsImpl(Task parentTask)
+        {
+            this.parentTask = parentTask;
+            this.application = parentTask.getApplication();
+        }
+
+        public String getQueueName()
```
Comment (dermot-hardy) [inline worker-document/src/main/java/com/hpe/caf/worker/document/tasks/AbstractTask.java]: `@Override` annotation missing.

File: worker-document/src/main/java/com/hpe/caf/worker/document/tasks/AbstractTask.java [hunk 26]
```diff
@@ -93,4 +104,58 @@ public final WorkerResponse handleGeneralFailure(final Throwable failure)
...
+        {
+            this.queueName = queueName;
+        }
+
+        /**
+         * Gets the custom data.
+         *
+         * @return The custom data map.
+         */
+        public Map<String, String> getCustomData()
```
Comment (dermot-hardy) [inline worker-document/src/main/java/com/hpe/caf/worker/document/tasks/AbstractTask.java]: `@Override` annotation missing.

File: worker-document-interface/src/main/java/com/hpe/caf/worker/document/model/ResponseOptions.java [hunk 27]
```diff
@@ -0,0 +1,55 @@
...
+     * Sets the custom data.
+     * @param customData Custom data which contains additional information for a document worker task.
+     */
+    void setCustomData(Map<String, String> customData);
+
+    /**
+     * Returns a {@link Task} that is associated with this object.
+     * @return a {@link Task} that is associated with this object.
+     */
+    Task getTask();
```
Comment (dermot-hardy) [inline worker-document-interface/src/main/java/com/hpe/caf/worker/document/model/ResponseOptions.java]: `@Nonnull` annotation missing.  This object should always have a task parent.

File: worker-document/src/main/java/com/hpe/caf/worker/document/JavaScriptDocumentPostProcessor.java [hunk 28]
```diff
@@ -0,0 +1,60 @@
...
+/**
+ * Implementation of {@link DocumentPostProcessor} which executes a Java Script for post-processing.
+ */
+public class JavaScriptDocumentPostProcessor implements DocumentPostProcessor
+{
+    private static final Logger LOG = LoggerFactory.getLogger(JavaScriptDocumentPostProcessor.class);
+
+    private final String postProcessingScript;
+
+    public JavaScriptDocumentPostProcessor(String postProcessingScript)
```
Comment (dermot-hardy) [inline worker-document/src/main/java/com/hpe/caf/worker/document/JavaScriptDocumentPostProcessor.java]: Parameter should be `final`.

File: worker-document/src/main/java/com/hpe/caf/worker/document/JavaScriptDocumentPostProcessor.java [hunk 29]
```diff
@@ -0,0 +1,60 @@
...
+    {
+        this.postProcessingScript = postProcessingScript;
+    }
+
+    @Override
+    public void postProcessDocument(final Document document) throws PostProcessingFailedException
+    {
+        LOG.trace("Executing post-processing script.");
+        final ScriptEngineManager engineManager = new ScriptEngineManager();
```
Comment (dermot-hardy) [inline worker-document/src/main/java/com/hpe/caf/worker/document/JavaScriptDocumentPostProcessor.java:45]: I think there might be a performance benefit to sharing the `ScriptEngineManager` rather than reconstructing it each time as here.  That said it is probably a premature optimisation to make without evidence of an issue.

**Review by dermot-hardy — CHANGES_REQUESTED** [review]

File: worker-document-testing-unit/src/main/java/com/hpe/caf/worker/document/testing/DocumentBuilder.java (hunk 17 above)
Comment (kploch) [reply worker-document-testing-unit/src/main/java/com/hpe/caf/worker/document/testing/DocumentBuilder.java]: First of all - testability. I prefer dependency injection instead of hardcoding the implementation in the constructor.
You might want to a mocked up implementation (or a coded testing implementation)

File: worker-document/pom.xml (hunk 12 above)
Comment (kploch) [reply worker-document/pom.xml]: I'll be honest - I hate that java itself lacks this type of features. But anyway, I changed that.

File: worker-document/src/main/java/com/hpe/caf/worker/document/DocumentMessageProcessor.java (hunk 10 above)
Comment (kploch) [reply worker-document/src/main/java/com/hpe/caf/worker/document/DocumentMessageProcessor.java]: Done

File: worker-document/src/main/java/com/hpe/caf/worker/document/DocumentPostProcessorFactory.java (hunk 16 above)
Comment (kploch) [reply worker-document/src/main/java/com/hpe/caf/worker/document/DocumentPostProcessorFactory.java:39]: Yep, created one. It's in the epic. Don't want to change it now as we're going to park this work.

File: worker-document/src/main/java/com/hpe/caf/worker/document/DocumentPostProcessorFactory.java (hunk 15 above)
Comment (kploch) [reply worker-document/src/main/java/com/hpe/caf/worker/document/DocumentPostProcessorFactory.java]: Done

File: worker-document/src/main/java/com/hpe/caf/worker/document/DocumentPostProcessorFactory.java (hunk 14 above)
Comment (kploch) [reply worker-document/src/main/java/com/hpe/caf/worker/document/DocumentPostProcessorFactory.java]: Done

File: worker-document/src/main/java/com/hpe/caf/worker/document/DocumentPostProcessor.java (hunk 11 above)
Comment (kploch) [reply worker-document/src/main/java/com/hpe/caf/worker/document/DocumentPostProcessor.java]: Done

File: worker-document-shared/src/main/java/com/hpe/caf/worker/document/DocumentWorkerConstants.java (hunk 9 above)
Comment (kploch) [reply worker-document-shared/src/main/java/com/hpe/caf/worker/document/DocumentWorkerConstants.java:33]: As discussed - this will probably apply to other task as well. We can discuss it as a follow up task.

File: worker-document/src/main/java/com/hpe/caf/worker/document/DocumentWorkerFactoryProvider.java (hunk 13 above)
Comment (kploch) [reply worker-document/src/main/java/com/hpe/caf/worker/document/DocumentWorkerFactoryProvider.java]: Done

File: worker-document/src/main/java/com/hpe/caf/worker/document/impl/ApplicationImpl.java (hunk 19 above)
Comment (kploch) [reply worker-document/src/main/java/com/hpe/caf/worker/document/impl/ApplicationImpl.java]: Done

File: worker-document/src/main/java/com/hpe/caf/worker/document/tasks/AbstractTask.java (hunk 24 above)
Comment (kploch) [reply worker-document/src/main/java/com/hpe/caf/worker/document/tasks/AbstractTask.java]: Nothing outside this class creates it. This was the reasoning. I was thinking about it but then decided I prefer this type of structure. I'm open to changing it if you give me advantage of doing so.

File: worker-document/src/main/java/com/hpe/caf/worker/document/tasks/AbstractTask.java (hunk 25 above)
Comment (kploch) [reply worker-document/src/main/java/com/hpe/caf/worker/document/tasks/AbstractTask.java]: Done

File: worker-document/src/main/java/com/hpe/caf/worker/document/tasks/AbstractTask.java (hunk 26 above)
Comment (kploch) [reply worker-document/src/main/java/com/hpe/caf/worker/document/tasks/AbstractTask.java]: Done

File: worker-document/src/main/java/com/hpe/caf/worker/document/tasks/DocumentTask.java (hunk 20 above)
Comment (kploch) [reply worker-document/src/main/java/com/hpe/caf/worker/document/tasks/DocumentTask.java]: I change the other class (app impl), but left this because I didn't want to update the interface. If you think it should be on the interface for Application, I'll change it.

File: worker-document/src/main/java/com/hpe/caf/worker/document/tasks/DocumentTask.java (hunk 22 above)
Comment (kploch) [reply worker-document/src/main/java/com/hpe/caf/worker/document/tasks/DocumentTask.java]: Done

File: worker-document/src/main/java/com/hpe/caf/worker/document/tasks/DocumentTask.java (hunk 23 above)
Comment (kploch) [reply worker-document/src/main/java/com/hpe/caf/worker/document/tasks/DocumentTask.java]: I understand your point but it's not really confusing. Let's not change it... Sorry. I can't get McA to change his code over again.

File: worker-document/src/main/java/com/hpe/caf/worker/document/tasks/DocumentTask.java (hunk 21 above)
Comment (kploch) [reply worker-document/src/main/java/com/hpe/caf/worker/document/tasks/DocumentTask.java]: Actually, this all fits into the margin, I changed it to be one line - hope this is ok. It doesn't mess up readability.
